src/GraphBLAS-sharp/MatrixAndVector.fs

```diff
@@ -37,32 +37,47 @@ and [<AbstractClass>] Vector<'a when 'a : struct and 'a : equality>() =
     abstract Length: int
     abstract AsArray: 'a[]
 
-    abstract Item: Mask1D<'a> -> Vector<'a> with get, set
-    abstract Item: int * int -> Scalar<'a> with get, set
-    abstract Item: Mask1D<'a> -> Scalar<'a> with set
+    abstract Item: Mask1D -> Vector<'a> with get, set
+    abstract Item: int -> Scalar<'a> with get, set
+    abstract Item: Mask1D -> Scalar<'a> with set
 
-    abstract Vxm: Matrix<'a> -> Mask1D<'a> -> Semiring<'a> -> Vector<'a>
-    abstract EWiseAdd: Vector<'a> -> Mask1D<'a> -> Semiring<'a> -> Vector<'a>
-    abstract EWiseMult: Vector<'a> -> Mask1D<'a> -> Semiring<'a> -> Vector<'a>
-    abstract Apply: Mask1D<'a> -> UnaryOp<'a, 'b> -> Vector<'b>
+    abstract Vxm: Matrix<'a> -> Mask1D -> Semiring<'a> -> Vector<'a>
+    abstract EWiseAdd: Vector<'a> -> Mask1D -> Semiring<'a> -> Vector<'a>
+    abstract EWiseMult: Vector<'a> -> Mask1D -> Semiring<'a> -> Vector<'a>
+    abstract Apply: Mask1D -> UnaryOp<'a, 'b> -> Vector<'b>
     abstract Reduce: Monoid<'a> -> Scalar<'a>
 
-    abstract EWiseAddInplace: Vector<'a> -> Mask1D<'a> -> Semiring<'a> -> unit
-    abstract EWiseMultInplace: Vector<'a> -> Mask1D<'a> -> Semiring<'a> -> unit
-    abstract ApplyInplace: Mask1D<'a> -> UnaryOp<'a, 'b> -> unit
+    abstract EWiseAddInplace: Vector<'a> -> Mask1D -> Semiring<'a> -> unit
+    abstract EWiseMultInplace: Vector<'a> -> Mask1D -> Semiring<'a> -> unit
+    abstract ApplyInplace: Mask1D -> UnaryOp<'a, 'b> -> unit
 
     static member inline (+) (x: Vector<'a>, y: Vector<'a>) = x.EWiseAdd y
     static member inline (*) (x: Vector<'a>, y: Vector<'a>) = x.EWiseMult y
     static member inline (+.*) (x: Vector<'a>, y: Matrix<'a>) = x.Vxm y
     static member inline (.+) (x: Vector<'a>, y: Vector<'a>) = x.EWiseAddInplace y
     static member inline (.*) (x: Vector<'a>, y: Vector<'a>) = x.EWiseMultInplace y
 
-and Mask1D<'a when 'a : struct and 'a : equality> =
-    | Mask1D of Vector<'a>
-    | Complemented1D of Vector<'a>
-    | None
+// and Mask1D<'a when 'a : struct and 'a : equality> =
+//     | Mask1D of Vector<'a>
+//     | Complemented1D of Vector<'a>
+//     | None
+
+// and Mask2D<'a when 'a : struct and 'a : equality> =
+//     | Mask2D of Matrix<'a>
+//     | Complemented2D of Matrix<'a>
+//     | None
+
+and [<AbstractClass>] Mask1D() =
```

Comment:
Why should these classes be abstract? I thought we were going to represent a single implementation for the mask, and it would be based on different implementations of the sparse matrix and vector. So, it seems to me, that using a mask should look like this: `currentVisited.VxmInplace matrix (Mask1D levels) BooleanSemiring.OrAnd`. This question requires discussion. Ultimately, in C API the raw matrix or vector is passed to the operations as mask.

Reply:
If we will only use sparse representations of the matrix and the vector (or the dense representation of the vector, but with an explicit indication of the semiring inside), will there be problems with the implementation of the mask as DU?